### ISSUE TITLE
v3 Roadmap

### ISSUE THREAD

**Issue by xxczaki** [node-fetch/node-fetch#668]
**Note: v3 beta is now available on npm under the `next` tag**

---

# v3 Roadmap

## General stuff

- [x] **BREAKING:** Drop support for Node.js 4 and Node.js 6 (#489, #624)
- [x] Update devDependencies (related to #489)
- [x] When users clone(), automatically create streams with custom highWaterMark (#386, PR open: #671)
- [x] Bundle TypeScript types (Support for TypeScript without need of installing `@types/node-fetch` package) (#669)
- [x] Replace Rollup with Microbundle (#643)
- [x] Introduce linting to the project (#303)
- [x] Simplify Travis CI build matrix (related to #431) - Dropping support for Node.js 4 & 6 will help
- [x] Response.statusText should not default (#578)
- [x] Use w3c defined message for AbortError (#657)
- [x] URL is not encoded as UTF-8 (#245)
- Deprecate timeout in favor of AbortController (#523)
- [x] Data URI support (#659)
- [x] Drop the optional dependency on encoding and move charset detection into a separate package. (In progress: [Richienb/fetch-charset-detection](https://github.com/Richienb/fetch-charset-detection) (PR Open: #694))
- [x] Drop existing blob implementation code and use fetch-blob as dependency instead (https://github.com/bitinn/node-fetch/issues/668#issuecomment-529175627)
- [x] Fix stream piping (#309, PR open: #670)
- [x] Fix stream resolving (#642, #665)
- [x] Do not use constructor.name to check object (#667, PR open: #673)
- [x] Properly forward request errors while piping response stream
- [x] Convert `Content-Encoding` to lowercase (#661, PR open: #672)
- [x] More tests for data URI (https://github.com/bitinn/node-fetch/issues/668#issuecomment-529175096)
- [ ] Drop credentials before redirection (#274)
- [x] Modernise the code behind FetchError and AbortError (https://github.com/bitinn/node-fetch/issues/668#issuecomment-530845161)
- [x] Propagate size and timeout to cloned response (#664)
- [x] Replace the calls to `url.parse` and `url.resolve` with calls to `new URL` to use the new UTF-8 awareness and remove the `utf8` package.
    - [x] `url.parse` (https://github.com/node-fetch/node-fetch/pull/701)
	- [x] `url.resolve` (https://github.com/bitinn/node-fetch/commit/e3cd4d22c3f1871fb41b17035ac83b47ad57fdf6)
- [ ] Document types of fetch errors more thoroughly (#549)
- [ ] Should update host if there's a redirection (#570, PR Open: #596)
- [x] **High priority** doesn't correctly handle multiple `content-type` values (#783)

## Additional:

- [ ] HTTP/2 support (#342, PR open: #796)
- [ ] Plugin/middleware system (https://github.com/bitinn/node-fetch/issues/668#issuecomment-533544800, PR open: #677)
- ~~WebStream API support (#387)~~ - needs more discussion and standard.
- ~~Support trailers (#330)~~ - trailer support has been removed (whatwg/fetch#772).
- ~~Caching (#68)~~ - would likely require opinionated implementation.
- ~~Cookies (#94)~~ - would likely require opinionated implementation.

## Other

- [x] Enable Dependabot to automatically update dependencies if they pass tests and are within an appropriate version range.
- [x] A way to back/sponsor the project (#680)
- ~~Create a final v2 release before releasing v3~~
- [ ] Update upgrade guide
- [ ] Release v1.x version of [Richienb/fetch-charset-detection](https://github.com/Richienb/fetch-charset-detection) before release.
- [ ] Add this package to [DefinitelyTyped/DefinitelyTyped/master/notNeededPackages.json](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/notNeededPackages.json) since we now ship the typings.

---

cc @bitinn @jimmywarting @TimothyGu @jkantr @gr2m @Richienb

**Comment by Richienb** [comment]
@xxczaki We sorta need Data URI support as in https://fetch.spec.whatwg.org/#data-urls, implemented in https://github.com/bitinn/node-fetch/pull/659.

**Comment by Richienb** [comment]
@xxczaki I've migrated the TypeScript types in https://github.com/bitinn/node-fetch/pull/669.

**Comment by bitinn** [comment]
Going through some items:

> BREAKING: Drop support for Node.js 4 and Node.js 6 (#489, #624)
> Update devDependencies (related to #489)

definitely

> When users clone(), automatically create streams with custom highWaterMark (#386, PR open: #563)

probably don't want an API, clone() should probably just create both stream with larger highWaterMark so that isomorphic users don't run into issues that easily

> Bundle TypeScript types (Support for TypeScript without need of installing @types/node-fetch package)

as long as there are people to review and maintain this one.

> Rewrite to CommonJS & get rid of Rollup & Babel (#643)

I would love to, but need to think about potential impact?

> Introduce linting to the project (#303)

I am fine with or without it, make rules more relax at first would be good.

> Simplify Travis CI build matrix (related to #431) - Dropping support for Node.js 4 & 6 will help

definitely

> Body conversion (#210)

jimmy can probably help you on that

> Response.statusText should not default (#578)

yes, follow the spec on this

> Use w3c defined message for AbortError (#657)

we should probably change to match browsers, but need to review how many error types we will end up with (quite a few, I reckon, in #549); error handling in browsers have always been a headache, and AFAIK Fetch just couldn't address it due to browser legacy and conventions.

> URL is not encoded as UTF-8 (#245)

With older node dropped this should be easy to do

> Deprecate timeout in favor of AbortController (#523)

I am at least a bit skeptical on this: `AbortController` sure gives a lot of control but `timeout`'s ease of use seems hard to beat. I would think providing a helper module (outside of node-fetch, using AbortController) that emulate `timeout` is needed before we just drop `timeout`.

**Comment by Richienb** [comment]
@bitinn I can be a backer code reviewer. There are heaps of things in the roadmap that needs doing and I'm up for creating and reviewing PRs to resolve them. TLDR: Please make me a collaborator. 👍

**Comment by bitinn** [comment]
@Richienb sure, added! make sure you ask other for review before merging risky changes.

Also we should decide on whether we merge changes on `3.x` branch or `master` branch; I was meaning to use `3.x`  but if we are already working on v3 release then `master` is fine too.

**Comment by Richienb** [comment]
@bitinn At the moment, I'm creating PRs against 3.x

**Comment by bitinn** [comment]
@Richienb let's go with `3.x` then, unless @xxczaki or @jimmywarting has any objections

**Comment by Richienb** [comment]
@bitinn Sure. I've also just created the Discord server which I'll finish setting up within the next hour: https://discord.gg/hh8ZSw7

**Comment by bitinn** [comment]
> Caching
> Cookies

I wasn't planning to tackle these within `node-fetch`, mostly because I would like to make these features possible by having a simple plugin system, it was discussed in an old issue.

> HTTP/2 support

I was waiting for node to provide some abstraction layer between HTTP/1 and HTTP/2 so that we can build upon it. Feels like a lot of work and a headache to maintain.

> WebStream API

I am quite interested in this, but the ecosystem doesn't seem to be ready, can we easily convert between Node Stream and WebStream nowadays? Can we hide such conversion so that users' code is truly isomorphic? Lots of things to consider.

I was planning to allow users to provide a custom "Stream" implementation, much like a custom `Blob` or `FormData`, but that feels too much...

**Comment by Richienb** [comment]
@bitinn How are you envisioning the simple plugin system to work?

For WebStreams, wouldn't https://www.npmjs.com/package/node-web-streams cut the cake?

**Comment by Richienb** [comment]
@bitinn In the Discord server, do we need a feed holding all of the activity in this repository? If so, add a webhook for all events to `https://discordapp.com/api/webhooks/619919240027308042/YpxDdMxaQ3oOgtoZfkAWovJ3A4pztZFWBD1pF0mU6rlQpz0vaRa-J1_sPMIHO24c0YtX/github` in the repo settings.

**Comment by xxczaki** [comment]
@bitinn 

```
probably don't want an API, clone() should probably just create both stream with larger highWaterMark so that isomorphic users don't run into issues that easily
```
I agree. It's also worth mentioning, that there is a [80$ bounty on this issue](https://github.com/sindresorhus/ky-universal/issues/8#issuecomment-513632746)

```
as long as there are people to review and maintain this one.
```
I'm happy to maintain types, so that shouldn't be a problem :smile:\

```
I am fine with or without it, make rules more relax at first would be good.
```
Sure, we can use ESLint wrapper, like [xo](https://github.com/xojs/xo) to get some reasonable defaults and turn off rules we don't like.

---

Let's leave HTTP2 support & WebStream API and wait for them to be fully implemented in Node.js Implementing HTTP/2 would indeed be a hard task (see for example https://github.com/sindresorhus/got/issues/167).

**Comment by bitinn** [comment]
@Richienb I haven't thought about it too much to be honest, maybe we don't need plugins at all, just allow users to set a `fetch.CookieJar` might be enough. AFAIK only caching and cookie require such external dependencies.

**Comment by Richienb** [comment]
@xxczaki 

> I'm happy to maintain types, so that shouldn't be a problem 😄\

In that case, we can go ahead and merge https://github.com/bitinn/node-fetch/pull/669

**Comment by xxczaki** [comment]
@Richienb Done :smile:

**Comment by Richienb** [comment]
@bitinn @xxczaki I'm currently working on adding the xo linter to the project. What type of spacing are we going for? Spaces? Tabs?

For now, I'll go with 4 spaces.

**Comment by xxczaki** [comment]
@Richienb Tabs are my favorite

**Comment by xxczaki** [comment]
@bitinn To be honest with you, I think it will be better if we stay with Babel (of course with updated & simplified configuration) & write ES6 code. However, instead of manually configuring Rollup, maybe we could use [microbundle](https://github.com/developit/microbundle)? It uses Rollup underneath, has built-in compression and doesn't require any configuration at all :smile:

**Comment by xxczaki** [comment]
@Richienb I will start working on:

- Rewrite to CommonJS & get rid of Rollup & Babel 
- Drop support for Node.js 4 and Node.js 6
- Update devDependencies

As soon as you add xo :smile:

**Comment by Richienb** [comment]
@xxczaki It turns out, not using a linter for 4 years may not have been a good choice. XO is spitting out 50 or so errors that can't be automatically fixed and I'm slowly grinding through them, adding ignore rules and fixing the fixable.

**Comment by gr2m** [comment]
I would suggest to add an automated release process to the list, based on 3 commit message conventions (`fix: ...` & `feat: ...` in commit subject, `BREAKING CHANGE: ...` in body). I would not be able to keep as many projects up-do-date as I do without the combination of @Greenkeeperio & @semantic-release.

I can help with both Greenkeeper & semantic-release setup and if there are any problems that would arise from it in future. I worked on Greenkeeper in the past and I am a maintainer at semantic-release myself.

**Comment by gr2m** [comment]
Regarding

> Introduce linting to the project

I would strongly suggest to use a zero-config linter, such as https://standardjs.com/ or https://prettier.io/ (without custom config). Nobody will ever be happy with any linting setup, but at least we can avoid having discussions about what linting rule to add or change, and focus on the code instead.

**Comment by xxczaki** [comment]
```
@xxczaki It turns out, not using a linter for 4 years may not have been a good choice. XO is spitting out 50 or so errors that can't be automatically fixed and I'm slowly grinding through them, adding ignore rules and fixing the fixable.
```

I will be happy to help with some of the errors, if you want to :smile:

**Comment by xxczaki** [comment]
> I would suggest to add an automated release process to the list, based on 3 commit message conventions (`fix: ...` & `feat: ...` in commit subject, `BREAKING CHANGE: ...` in body). I would not be able to keep as many projects up-do-date as I do without the combination of @greenkeeperio & @semantic-release.
> 
> I can help with both Greenkeeper & semantic-release setup and if there are any problems that would arise from it in future. I worked on Greenkeeper in the past and I am a maintainer at semantic-release myself.

I will need to think about that, semantic-release seems really nice (I've actually used it in one of my projects in the past), but I'm not sure about Greenkeeper... Have you ever used [Dependabot](https://dependabot.com/)? It has some additional features and was also recently aquired by Github, so it might be a better option.

**Comment by gr2m** [comment]
I’ve always used greenkeeper due to my affiliation to it, but in the end they all do the same, so I really don’t mind which we end up using, as long as we have one. As we already have 100% test coverage we will be able to merge dependency updates quickly and with confidence, no matter the service that sends the PRs.

**Comment by Richienb** [comment]
I'd say we use Dependabot since it serves a similar purpose and is run by GitHub.

**Comment by gr2m** [comment]
@bitinn have you signed up for GitHub Actions yet? It would be good to use it for semantic-release and probably other CI / automation in future. It's really great. Let me know if you did but are waiting for approval

**Comment by Richienb** [comment]
@xxczaki I've added linting in https://github.com/bitinn/node-fetch/commit/43b2aa4cfab3bad79e87d858d97f0eb94c9f38af. You can run `npm run lint` or `yarn lint` to see the errors.

**Comment by Richienb** [comment]
@xxczaki Are we going to be using microbundle or regular commonjs?

**Comment by xxczaki** [comment]
@Richienb Let's use microbundle :smile:

In terms of linter, I resolved most of the errors. The only left are the ones related to deprecated `url.parse` and `url.resolve` API ;)

**Comment by xxczaki** [comment]
Note: Let's hold up from migrating to WHATWG `new URL()` API, as this issue is still open: https://github.com/nodejs/node/issues/12682

**Comment by Richienb** [comment]
@xxczaki microbundle is giving the warning: `Using named and default exports together. Consumers of your bundle will have to use fetch['default'] to access the default export, which may not be what you want.` which I think means instead of doing `const fetch = require("fetch")` people would have to do `const {default} = require("fetch")`. I'll push the changes so you can take a look.

**Comment by xxczaki** [comment]
@Richienb Well, that's not ok. We need to do something about it; I will try to take a look.

**Comment by xxczaki** [comment]
> @xxczaki microbundle is giving the warning: `Using named and default exports together. Consumers of your bundle will have to use fetch['default'] to access the default export, which may not be what you want.` which I think means instead of doing `const fetch = require("fetch")` people would have to do `const {default} = require("fetch")`. I'll push the changes so you can take a look.

I don't know what's the best way to fix this issue. We could use only named exports:

```js
export {
	fetch,
	Headers,
	Request,
	Response,
	FetchError
};
```

But that would require users to do `const {fetch} = require('node-fetch')`, which is not ideal. We would also need to rewrote some tests. 

Maybe I will come up with something soon and if not, we would either keep the old Rollup config (but with updated Babel 7 configuration) or rewrote everything to CommonJS.

**Comment by Richienb** [comment]
> Maybe I will come up with something soon and if not, we would either keep the old Rollup config (but with updated Babel 7 configuration) or rewrote everything to CommonJS.

Rewriting in CommonJS isn't really an option since the issue raised by microbundle only affects the converted CommonJS code.

**Comment by xxczaki** [comment]
We can try a solution from https://github.com/developit/microbundle/issues/306#issuecomment-459103476.

Also @gr2m [suggested](https://github.com/bitinn/node-fetch/issues/643#issuecomment-494496251) usage of `@pika/pack`, we might also consider that as an option.

**Comment by Richienb** [comment]
@xxczaki LGTM. We can try that.

**Comment by bitinn** [comment]
Commenting on 2 things:

- When closing an issue, would be better to include a link to relevant commit for original author to review.

- Recent data URI support PR introduced something `node-fetch` hasn't supported until now: **non-http url scheme**. I think it's worth commenting the code to address this change. (and could use more test to make sure new code is largely spec-compliant)

https://fetch.spec.whatwg.org/#scheme-fetch
https://fetch.spec.whatwg.org/#data-url-processor

**Comment by bitinn** [comment]
There are 2 more breaking change I have been meaning to do in v3, see if you can get to it:

- **Drop the optional dependency on `encoding` and move charset detection into a separate package.** The charset conversion isn't a part of Fetch spec, but a legacy extension. By doing this we should solve many users' bundling problem when using `webpack` (it doesn't like optional require).

I was planning to do this but never got around, you can open a new repo for the separate package, or I could add you to my empty repo:

https://github.com/bitinn/fetch-charset-detection

- **Drop existing blob implementation code and use `fetch-blob` as dependency instead.* The reason for this is we would like to support other compatible blob implementation in the wild. And `fetch-blob` is just extracted from our implementation and with proper tests.

This one I have already done, but could use people's help maintaining it:

https://github.com/bitinn/fetch-blob

**Comment by Richienb** [comment]
@bitinn Should the `fetch-blob` code be transpiled directly into the final output or as an extra dependency?

**Comment by bitinn** [comment]
@Richienb just as a dependency in package.json, unless dropping our built-in blob break other things I am not aware of.

**Comment by Richienb** [comment]
@bitinn In the `node-fetch` implementation, the `BUFFER` constant is exported whilst in `fetch-blob` it isn't.

**Comment by Richienb** [comment]
@bitinn For automated dependency updates, we've decided that it's best to use [Dependabot](https://dependabot.com/) set to auto-merge for in-range and build-passing updates. However, the final choice is up to you.

**Comment by bitinn** [comment]
I would say I don’t know nodejs bundling and workflow enough these days to judge whether one choice is superior over another, hopefully other members could offer their opinion.

(From my phone)

**Comment by bitinn** [comment]
Another TODO:

For years, we have been getting issues like #309, #642, #665; and I think the root cause comes down to these 2 problems:

1. We did not properly forward request errors while piping response stream:

https://github.com/bitinn/node-fetch/blob/master/src/index.js#L195

we should investigate using `pump` module or `stream.pipeline` (in node v10+) to address #309 (note that decompression also involve piping and need to be handled as well.)

https://nodejs.org/es/docs/guides/backpressuring-in-streams/#the-problem-with-data-handling

2. After receiving the response and consuming it, we resolve on response stream's `end` event:

https://github.com/bitinn/node-fetch/blob/master/src/body.js#L259

This is apparently not safe in some rare situations: we got all the data the remote server can send, but the underlying request is still ongoing, we can't resolve now, we need to wait for response `close` event to be safe to resolve, otherwise we need to reject it. This should solve #642 and #665 

https://nodejs.org/api/http.html#http_http_request_url_options_callback

Let me know if anything is unclear.

**Comment by jimmywarting** [comment]
> @bitinn just allow users to set a `fetch.CookieJar`

Not sure if that would be enough. wouldn't it create some conflicts when using it that way?

maybe would want something like this:

```js
fetch = new Fetch({
  cookieJar,
  base: 'https://httpbin.org',
  redirect: 20,
  plugins: [ ... ],
  onFetch: fn // similar to service worker onfetch (intercepting)
})

fetch('/get') // https://httbin.org/get
```

**Comment by Richienb** [comment]
@bitinn @xxczaki Do we *need* to rely on our version of Blob instead of using other implementations such as the [`blob`](https://www.npmjs.com/package/blob) package?

**Comment by bitinn** [comment]
I have fetch-blob package ready to return a response as Blob, and we expect users might pass in any Blob implementation as body. Hope this answer your question.

(From my phone)

**Comment by aksel** [comment]
Re: plugin system for cookies and caching, would a middleware-like approach make sense?

Something like so:
```
const cookieJar = new Jar();
const cache = new MemoryCache();

fetch.before((request) => {
    if (cache.hasCachedResponse(request)) {
        return cache.getCachedResponse(request);
    }

    const requestWithCookie = cookieJar.applyCookies(request);

    return requestWithCookie;
});

fetch.after((request, response) => {
    cookieJar.store(request, response);
    cache.store(request, response);
});
```

**Comment by jimmywarting** [comment]
> Re: plugin system for cookies and caching, would a middleware-like approach make sense?

I would like it more to be service worker alike. ppl are already familiar with it. 
code could be written to work cross service-worker and node

```js
self.addEventListener('fetch', (event) => {
  event.respondWith(
    caches.match(event.request).then((resp) => {
      return resp || fetch(event.request).then((response) => {
        return caches.open('v1').then((cache) => {
          cache.put(event.request, response.clone());
          return response;
        });  
      });
    })
  );
});
```

**Comment by Richienb** [comment]
@bitinn This middleware system of `.before` and `.after` could be how the plugin system could work (https://github.com/bitinn/node-fetch/issues/668#issuecomment-529118130).

cc @xxczaki

**Comment by gr2m** [comment]
fyi I created a plugin API to hook into the request cycle for https://github.com/octokit/rest.js/ which is build upon node-fetch. I started out with before/after, but soon realized that it's not sufficient, there are valid usecases where you need to wrap the life cycle entirely. This is the hook library I created for it if anyone is interested: https://github.com/gr2m/before-after-hook/

I like @jimmywarting approach 👍 

I would be very careful to not add too much complexity to `node-fetch`. Hooking into request lifecycle also has the risk of memory leaks, or blowing up memory usage in generate, if you are not very careful. 

@bitinn [you mentioned above](https://github.com/bitinn/node-fetch/issues/668#issuecomment-529118130) that a plugin system for cookies & caching was discussed previously, could you point me to that discussion?

**Comment by bitinn** [comment]
I believe this is Timothy's original idea for plugins, not sure if it's still the preferred way:

https://github.com/bitinn/node-fetch/pull/229#issuecomment-283185660

**Comment by bitinn** [comment]
Basically, there are 2 types of problems a plugin system can solve:

A. problems with external dependencies, like form-data, blob, cookies etc.
B. problems with hooking or hijacking the requests, like capturing, debugging, or even cookies like above.

I think we need to decide on whether we would like a system that solve both A and B, or keep it simple and solve A first (I do see value in B, just not sure if it needs to be the same system?)

**Comment by bitinn** [comment]
As for @jimmywarting's solution, while using service worker API is nice, my concern is it's designed for caching, not sure it fits other purposes beyond cookies?

Also service-worker like solution is rare on nodejs, can't say people are comfortable with it, but then I haven't used service-worker myself, so I am biased.

**Comment by xxczaki** [comment]
First of all, sorry for not being too active recently, but I was at an event in Canada without a decent internet connection :smile:

---

I like @Richienb [solution](https://github.com/bitinn/node-fetch/pull/677/) (which AFAIK is inspired by [gr2m/before-after-hook](https://github.com/gr2m/before-after-hook/)). At first it looks simple, but I would still like to see some examples (cookies etc.).

Service Worker API is definitely modern, but is indeed still rare in Node.js ecosystem and might look complicated for some people.

**Comment by gr2m** [comment]
> I think we need to decide on whether we would like a system that solve both A and B, or keep it simple and solve A first (I do see value in B, just not sure if it needs to be the same system?)

I would prefer to fix A first. If possible, I would try to avoid a plugin architecture, the browser's `fetch` doesn't have that either. As a rule of thumb I would suggest: if you can accomplish what you want by wrapping `fetch`, then you should probably do that instead of adding request lifecycle hooks into `node-fetch` itself for convenience only

**Comment by bitinn** [comment]
About a newly added dependency on `utf8` package, is it necessary at all? I believe nodejs v8+ already has whatwg-url compatible object when you do `new URL()`, which should be utf-8 aware.

Please add to test to verify `utf8` is indeed necessary.

It was introduced via this commit: https://github.com/bitinn/node-fetch/commit/f3b00aaf8ed847fd3f30b1b7182cdafa411d02f0

**Comment by Richienb** [comment]
@bitinn It seems like we need to replace the now deprecated calls to `url.parse` and `url.resolve` with calls to `new URL` to use the new UTF-8 awareness.

**Comment by bitinn** [comment]
@Richienb yep looks like nodejs to pushing people towards `new URL`, we should do that.

**Comment by bitinn** [comment]
About the plugin system:

So far we already have (A) middlware, (B) hooks, (C) service worker approaches.

I agree with @gr2m on the hooks: people could use these hooks for a variety of things beyond just caching or cookie supports, once we add them, we are committing to support them for a long time, that's a non-spec burden we should be very cautious with. (Also see how long it took us to deprecate and remove non-spec charset detection...)

My personal criteria for a good solution:

- **Isomorphic / Non-intrusive**: people who want automatic *cookie* or *caching* support in `node-fetch`, often happen to want *isomorphism*, eg. write their code for both front and backend. In this case, the more we deviate from browser/spec, the more people are uncomfortable with it.

- **Allow for iterations**: we are probably to going to iterate on this solution; a low-impact, closed solution is better than a generic, open solution. The reason is people might start depending on our open solution, making iteration very difficult under the road.

So to me, a solution forward looks like:

- Pick our favorite cookie jar and cache implementation.
- Build `node-fetch` towards Fetch spec on cookie and caching, using an abstraction of their API.
- Then allow users to inject any compatible object that fits our abstraction, so that they can use any cookie jar or cache module underneath.

Yes, this looks very limiting, but:

- All users have to do is either: inject our preferred cookie jar and cache module, and everything behaves close to browsers.
- Alternatively, write a small adapter around their preferred cookie jar and cache module, inject that everything behaves close to browsers.
- Only the injection code itself is nodejs specific, anything else should be Fetch spec compliant and thus more or less isomorphic.
- If we ever changes this API, the users only have to update their injection code, which is also easier.
- If the users don't want cookie or caching, simply do nothing, and `node-fetch` behaves like v2, making upgrade from v2 to v3 easier.

Obviously this puts the work on us to find the cookie jar and cache system, then build the abstraction, then implement all those nitty-gritty according to Fetch spec, which is a lot of work. But I think it's better than non-spec API.

(By the way, there are quite a few popular package that build on top of node-fetch and add caching or cookie; [make-fetch-happen](https://www.npmjs.com/package/make-fetch-happen) is one, npm team forked `node-fetch` during v2, then wrap around it adding caching support, which `npm` now uses. Maybe there are solutions we can learn from.)

**Comment by xxczaki** [comment]
> Note: Let's hold up from migrating to WHATWG `new URL()` API, as this issue is still open: [nodejs/node#12682](https://github.com/nodejs/node/issues/12682)

cc @bitinn @Richienb

**Comment by bitinn** [comment]
I think we don't support relative url and should be fine?

**Comment by Richienb** [comment]
note to self: NodeJS v8 end of life is in 3 months.

**Comment by Richienb** [comment]
@bitinn @xxczaki A newly added unit test for data uris revealed a bug. node-fetch has a problem parsing `Hello World!` from `data:,Hello%20World!` whilst window.fetch in the browser handles it fine.

Build details: https://travis-ci.org/bitinn/node-fetch/jobs/588851048#L725-L732

**Comment by xxczaki** [comment]
> note to self: NodeJS v8 end of life is in 3 months. When to drop support?

At least a few months after it has been deprecated ;) While I don't like supporting old Node.js versions, I think (considering the large scale of node-fetch) that we should give developers more time for migration :smile:

**Comment by gr2m** [comment]
I'd drop Node 8 support the moment we have a breaking release anyway. As soon as node-fetch dependencies stop dropping support for it, there will be no alternative anyway. Stop supporting Node 8 as soon as it adds to maintenance overhead.

**Comment by bitinn** [comment]
The main pressure to support node v8 are probably from AWS Lambda users right? Where they are stuck with their environment?

(From my phone)

**Comment by aksel** [comment]
Lambda supports 10.x runtimes, as of May 2019.

**Comment by xxczaki** [comment]
@gr2m You are right, since `v3` going to be a major release, we can stop supporting Node.js 8 straight away.

@bitinn It looks like AWS Lambda [supports Node.js 10](https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html). Are there any other use cases, where Node.js 10 isn't supported (can't think of any right now)?

**Comment by Richienb** [comment]
@xxczaki @bitinn A recent addition to the unit tests have uncovered an issue with data URI parsing: https://travis-ci.org/bitinn/node-fetch/jobs/588851048#L455-L464

**Comment by Richienb** [comment]
Dropped NodeJS v8 in https://github.com/bitinn/node-fetch/commit/bf65e085a3e526d98dc4e03f797739d6b9c8276e

**Comment by xxczaki** [comment]
> A recent addition to the unit tests have uncovered an issue with data URI parsing: [travis-ci.org/bitinn/node-fetch/jobs/588851048#L455-L464](https://travis-ci.org/bitinn/node-fetch/jobs/588851048#L455-L464)


@Richienb I think this is an issue with `iconv` package (`encoding` relies on it). I tried replacing the `encoding` package with `iconv-lite`, which is much more lightweight, faster and most importantly - actively maintained. Unfortunately the issue still existed. I will file an issue and see what they say.

But since we are going to drop the optional dependency on `encoding` (see [this comment](https://github.com/bitinn/node-fetch/issues/668#issuecomment-529175627)), I don't think we need to worry much about it :smile:

**Comment by Richienb** [comment]
@xxczaki Are we going to be moving the whole conversion system or just for character detection to a new library?

**Comment by xxczaki** [comment]
@Richienb I don’t have a strong opinion on this one.

cc @bitinn

**Comment by Richienb** [comment]
There's a bug in body parsing in which the detection regexes are not strict enough. For instance ``<metaV->?`8{j=E-oIA`TaT:ELJS1d4n|y$4>5|wH?N.}tRKq0"ueb]b<ac%@lfy_\BG;%Kc{Y*Hzij-charset='IX CdWIPMj][/nN4IgY3}a2H\h/fs/a<15@PUxtI'`` passes the regex.

I'll fix that with proper AST parsing in the new package which will unfortunately, use more dependencies.

**Comment by bitinn** [comment]
You mean charset detection and encoding conversion? Body parsing shouldn’t be a part of v3 at all, leave it to a separate package.

(From my phone)

**Comment by Richienb** [comment]
@bitinn In that case, what would the package be called?

**Comment by bitinn** [comment]
@Richienb https://github.com/bitinn/fetch-charset-detection

I haven't started working on it, also unlike `fetch-blob`, it shouldn't be a dependency of `node-fetch`, the user should just pass this package a stream or string (eg. from `res.body` or `res.text()`) and it should just do the encoding conversion to UTF-8.

That's my plan anyway.

**Comment by Richienb** [comment]
@bitinn I've started development on my local machine. Here's the link if you have VSC live share: https://prod.liveshare.vsengsaas.visualstudio.com/join?D21F822B9DE89BE197D6A17E1037CC89978F

**Comment by bitinn** [comment]
@Richienb also, feel free to open your own GitHub repo and npm package, if submitting PR and waiting for me to merge and publish new release is too slow in this case.

**Comment by Richienb** [comment]
`cough` `cough`

**Comment by Richienb** [comment]
I've noticed that in `cross-fetch` if the request URL starts with `//` such as `//google.com`, it is automatically prepended with `https:`: [`lquixada/cross-fetch/src/node-ponyfill.js#L7-L9`](https://github.com/lquixada/cross-fetch/blob/master/src/node-ponyfill.js#L7-L9)

I'm thinking that we should do something similar here. Since NodeJS doesn't run on a protocol, we could assume `http` if none is provided.

**Comment by jimmywarting** [comment]
> I'm thinking that we should do something similar here. Since NodeJS doesn't run on a protocol, we could assume http if none is provided.

if we would do that, would it be crazy to default to https? More and more is being pushed towards secure by default to prevent spying eyes. There is popular addons like [HTTPS Everywhere](https://www.eff.org/https-everywhere) if you want to be explicit then you have to write `http://`

**A reason for not allowing relative protocol**

> As the web moves towards [HTTPS everywhere](https://w3ctag.github.io/web-https/#h-motivating-a-secure-web), the use of protocol-relative URLs has become an [anti-pattern](https://www.paulirish.com/2010/the-protocol-relative-url/), exposing some sites to man in the middle compromises and is therefore best avoided. <br> https://webhint.io/docs/user-guide/hints/hint-no-protocol-relative-urls/

Omitting the http(s): is just a laziness so the node-fetch needs to do more work... i'm afraid that if we allow `//example.com` then your code might end up in the web and do different things depending on your protocol. i think it's almost better to be explicit about it.

**Comment by Richienb** [comment]
@jimmywarting In that case, should we default to `https` or throw an exception?

**Comment by jimmywarting** [comment]
we are already throwing and saying that we don't allow relative urls...?

**Comment by Richienb** [comment]
@jimmywarting Oh right. 👌

**Comment by itskibo** [comment]
What about #49? This is still an ongoing issue, and while the workaround does work, it's a hassle to manually keep track of credentials through HttpOnly cookies.

**Comment by xxczaki** [comment]
As [Hacktoberfest](https://hacktoberfest.digitalocean.com) is ongoing, we can add the `Hacktoberfest` label to encourage people to contribute :smile:

cc @Richienb @bitinn @jimmywarting

**Comment by Richienb** [comment]
@xxczaki Sounds good to me!

**Comment by xxczaki** [comment]
You can also pin the `Hacktoberfest` labeled issue, so that they will appear on top.

**Comment by isaacs** [comment]
Getting back to the subject of cookie jars and caching, I think the current lack of a plugin/middleware/hook architecture is actually _totally fine_, and this module should stick to following the specification to the letter.  To the extent that there is a call for a specific plugin, it's better (as @bitinn suggests) to limit the extension to the smallest possible surface area, and strictly define what that specific kind of plugin would need to do.

`make-fetch-happen` wraps node-fetch (or in newer versions, minipass-fetch, which is almost identical) and adds caching with `cacache` and `http-cache-semantics`.  While it would certainly be _easier_ if fetch had some facility to plug something in, a general purpose plugin system would (I think) be overkill in the extreme.

We haven't added a cookiejar to make-fetch-happen, for the simple reason that npm doesn't use cookies when it talks to the registry.  (For interesting/boring historical reasons, it uses a bearer token sent in a request headers.  If I had it all to do over again, I'd use a cookie, since it's the tool for that job.)

Cacache is a general-purpose file system content-addressable cache.  The main point of complexity in `make-fetch-happen` is getting all the `http-cache-semantics` stuff just right.  Since that _is_ a browser specification, it would be good for fetch to implement _that_ part, and define a contract for how the cache provider would work (or just use cacache; it's pretty good, and very well play-tested at this point).  Then the only spec deviation might be something like `fetch(url, { cacheFolder: './cache' })`.

Same with cookies.  There's some browser-specified logic about when a cookie should be sent or not, and it would be good for fetch to implement that to the spec.  But even in that case, better to just pick a cookiejar implementation, and say "version x.y.z supports cookies now, you're welcome".

I don't think it's very likely that anyone wants a cache or cookiejar implementation that _doesn't_ match the browser behavior.  When there are 3 compelling plugin/hook/middleware use cases that have divergent behavior needs, it may be worth considering.  But it is practically a guarantee that a plugin/middleware system will leak internal implementation details that users will come to rely on, and it'll end up being a source of regrets.

**Comment by xxczaki** [comment]
I'm trying to work out the way to fix #642, based on what David wrote (https://github.com/bitinn/node-fetch/issues/668#issuecomment-529210393). Replacing the `end` event with `close` simply won't work, as it will cause timeout (nothing has closed the stream source, so the `close` event will never be emitted). I don't really know how to handle this :confused: Manually closing a stream using a `stream.destroy()` function probably isn't what we are looking for.

Any ideas?

cc @bitinn @Richienb

**Comment by isaacs** [comment]
@xxczaki I'd probably just document it as a known edge case.  Node's core streams are excessively async, so the socket close event doesn't come until some time after the end of the data.  The `res.text()` promise should end as soon as it has its data, and forcibly destroying the stream is almost certainly going to do more harm than good.

**Comment by bitinn** [comment]
@xxczaki after re-reading the 3 original issues, I think by using `pump` we have solved #309 and #665. As for #642, I agree with isaacs, we probably shouldn't change our resolve-on-stream-end design. Instead we could add some kind of verbose mode in future, that log something on response body stream `close` for debug usage.

This is obviously a feature extension to Fetch, so we are in more rush. And I am not aware of Web Stream Spec mentioning a similar distinction between `end` and `close`. They have a reader and that's about that. So I think no one is going to come knocking for this one.

https://streams.spec.whatwg.org/#example-manual-read

**Comment by Richienb** [comment]
At the moment [`fetch-charset-detection`](https://github.com/Richienb/fetch-charset-detection) has implemented support for body conversion, content type extraction, getting the total bytes and writing to a stream. Is there anything else that needs to be implemented?

**Comment by xxczaki** [comment]
@Richienb What about that `Fix main function getting named exported as fetch["default"]`? Is it just a warning, that can be ignored or something we need to worry about?

We also need to replace deprecated `url.parse()` with WHATWG `new URL()`. That will take care of utf-8 url encoding, so we will be able to remove the `utf8` dependency.

**Comment by Richienb** [comment]
@xxczaki The TODO item I was referring to was *Drop the optional dependency on encoding and move charset detection into a separate package*. Do we need to transfer any other functions there or is it ready for release?

**Comment by xxczaki** [comment]
@Richienb Oh, ok 😄 I think it is ready for release, but let’s wait for David, since he suggested this change ;)

cc @bitinn

**Comment by bitinn** [comment]
@Richienb I think https://github.com/Richienb/fetch-charset-detection could use a doc update and migrating existing tests from this repo.

(If you don't want to maintain a separate doc, we could update our v3 README, provide some code examples using this new package.)

After that, we should drop the `encoding` dependency and the relevant code from our v3 codebase.

@xxczaki I haven't read the Upgrade doc, if it's been updated, then we are ready for a v3 alpha release.

Can we also remove the `utf8` package dependency introduced earlier? It looks like we are dropping node v8 support as well, there shouldn't be any issue using `new URL()`.

**Comment by Richienb** [comment]
@bitinn The documentation for the charset detection is automatically generated so that shouldn't be a problem. For now, I'll release an initial 0.x version and I'll update it to 1.x before the official release of Node Fetch v3.

Before merging the v3 branch into master, please publish a final update for v2.

**Comment by bitinn** [comment]
@Richienb @xxczaki I have read every commit on `master` after 2.6.0 release, I do not believe it's a good idea to release 2.6.1 or 2.7.0, because:

- If the data-url support faces any issue, we would rather fix it in v3, not v2.
- If the meta tag charset detection has issue, we would rather fix it in `fetch-charset-detection`.
- If the error message isn't quite as verbose, v3 should cover them.
- If the `browser` export isn't quite as nice, well, we have dropped them in v3.
- If the `readme` has issues, we would certainly fix them in v3.

So to me these changes are all v3 related, not v2 critical, releasing them might even introduce more maintanence for us.

I could certainly copy `master` into `v2` branch for preservation purpose. But releasing v2 doesn't appear to be necessary. As we currently have no intention to backport bugfix on v2.

Thoughts?

**Comment by xxczaki** [comment]
@Richienb I think since [Richienb/fetch-charset-detection](https://github.com/Richienb/fetch-charset-detection) is available, we can remove the optional dependency on `encoding` and it's code references.

**Comment by Richienb** [comment]
@bitinn In that case, copy the current `master` to `v2` and then replace `master` with `v3`.

@xxczaki I'll remove the `encoding` package in a future PR adding support for fetch-charset-detection.

**Comment by bitinn** [comment]
@Richienb @xxczaki  `2.x` branch is now up-to-date with `master`, and the same branch protection as `1.x`.

**Comment by Richienb** [comment]
@bitinn Since we're developing on the 3.x branch, I think it may be beneficial to set it as the default branch or move its contents to master.

**Comment by bitinn** [comment]
@Richienb please fix the `v3` branch build status first.

Also it's not possible to replace the `master` branch with `v3` without breaking history for other users, nor is setting `v3` branch as main branch a good option.

The best way forward is to fix the `v3` build status, then merge `master`, resolve whatever conflicts there is, then I can fast-forward the `v3` result back into `master`.

**Comment by Richienb** [comment]
Something changed in NodeJS v13.5 where the builds that were passing in v13.4 are now failing.

**Comment by bitinn** [comment]
@Richienb possible cause: this change, which trigger `error` event more than once, so the test failed because we call `done()` more than once..

https://github.com/nodejs/node/pull/30869

Should be able to fix it by changing the test from `body.on` to `body.once`, just wondering if this will cause any issues (could log a upgrade guide to mention since we are using pipeline to forward error, the res.body error event might be triggered multiple times?)

**Comment by Richienb** [comment]
It worked! 🎉 

https://travis-ci.org/bitinn/node-fetch/builds/630233493

**Comment by Richienb** [comment]
Just some speculation: it might be a good idea to move this repo, `fetch-charset-detection` and `fetch-blob` all in the same org.

**Comment by bitinn** [comment]
@Richienb Sure thing, I also think it's time to create such an org. Want me to open one? Or you and @xxczaki can take the initiative, then add me as an admin later. I will transfer the project once you are done.

**Comment by bitinn** [comment]
Also if anyone is working on https://github.com/bitinn/node-fetch/pull/701#issuecomment-565800417, late me know, probably the last thing holding up our 3.x pre-release?

**Comment by Richienb** [comment]
@bitinn I'll set up the org.

**Comment by Richienb** [comment]
@bitinn I've set up the org and invited all of the `node-fetch` collaborators. [Accept your invite](https://github.com/node-fetch).

I'll transfer in `fetch-charset-detection` and I'll need you to do the same for `node-fetch` and `fetch-blob`.

**Comment by Richienb** [comment]
@bitinn ^^

**Comment by bitinn** [comment]
Welcome to the new `node-fetch` org everyone!

I am wondering: now that we are an org, can any admin setup GitHub sponsor for the repo/org?

Previously it was limited by me not having access to GitHub sponsor due to my citizenship.

**Comment by Richienb** [comment]
@bitinn That feature is in beta. In order to join the waitlist, we need an "organization bank account" which can't be substituted for a personal one. https://github.com/sponsors/node-fetch/waitlist

**Comment by bitinn** [comment]
@Richienb I guess someone can look into whether this is feasible; if not, we can still associate individual GitHub sponsor account with the repo's sponsor button, I have no problem with that.

**Comment by gr2m** [comment]
You can link to [our open collective](https://opencollective.com/node-fetch) which has no such requirement as Open Collective itself acts as the fiscal sponsor. But without a plan on what to do with the money it will be hard to raise any, from my experience.

**Comment by tjconcept** [comment]
Can I ask, what was the motivation for removing the `browser` field in `package.json`?

It seems very cheap and actually works (for me) in all situations so far. I don't wan't to use e.g. `cross-fetch` as it bundles a polyfill. It would also mean relying on their release schedule.

**Comment by Richienb** [comment]
@tjconcept

`cross-fetch` can still be imported as a drop-in replacement for `node-fetch` for browser compatibility. If you want to use `node-fetch`, tools like Browserify still have to go in and polyfill `http`, `https`, `url`, `zlib` and `stream` which are all Node libraries which would result in large bundle sizes. On the other hand, `cross-fetch` would route to `whatwg-fetch` which uses `XMLHttpRequest` - an API that has existed since the dawn of IE and never has to be polyfilled.

**Comment by tjconcept** [comment]
> `cross-fetch` can still be imported as a drop-in replacement for `node-fetch` for browser compatibility.

Not without increasing the bundle size significantly due to `whatwg-fetch`.

> If you want to use `node-fetch`, tools like Browserify still have to go in and polyfill

That's not what happens in practice. When I bundle `node-fetch@2` using Browserify, I get just the `browser.js` file bundled.

**Comment by bitinn** [comment]
The truth is node-fetch is not designed to work with browsers. You should not require node-fetch for browser build. I don’t know why people want to do it as we are not providing much value to browser here.

And you can quite literally write your own module that expose the same browser field if you prefer that. In fact the existence of many modules, including cross-fetch and isomorphic-fetch, that do things just a bit different from each other means it’s literally impossible for us to cover all use cases.

(From my phone)

**Comment by tjconcept** [comment]
> The truth is node-fetch is not designed to work with browsers.

Absolutely not, and I don't want to use it for browsers. I consider it a Node.js specific polyfill. I just want it to stay out of browser bundles.
The new approach means it will end up in more browsers (accidentally or not).

> write your own module

Obviously. Or use one of the existing ones. I'll need to do that, of course, but I'm just asking for the motivation because I don't see the reason to.
At least, I think you should provide an upgrade path with similar behaviour (`cross-fetch` has not, as pointed out above).

**Comment by gr2m** [comment]
Another reason not to add it is that the `"browser"` package.json field has never been standardized. It's causing a lot of friction for module authors. Only recently the Node modules group agreed on a way forward, which all the bundlers have to catch up with first: https://nodejs.org/api/esm.html#esm_conditional_exports

It's easy for you to add the "browser" field in your app and then migrate to something else, it's much harder to do the same for a widely used project such as node-fetch.

@tjconcept if you want to further discuss this topic, can you please create a separate issue? A lot f people subscribed to this issue to stay up to date with latest news on v3, I'd like to reduce the noise for them.

**Comment by jimmywarting** [comment]
`node --require polyfill.js main.js` maybe will help some ppl to add stuff to `globalThis` in there node environment?

**Comment by dobromyslov** [comment]
Any plans in future to deprecate this lib like the request? Is node-fetch a good substitution for the request lib?

**Comment by csvan** [comment]
@dobromyslov love how you tell people to migrate to `node-fetch` over in https://github.com/request/request/issues/3142#issuecomment-586906981 without even knowing if this lib is a "good substitution" or even actively maintained. Pathetic.

**Comment by dobromyslov** [comment]
Meanwhile migrated from the request lib to node-fetch. Nothing difficult. Just took me some extra time to replace request.qs with code based on URLSearchParams, correctly convert nested query params, correct form params and JSON handling. And im glad node-fetch has Promises out of the box. Profit.

**Comment by jimmywarting** [comment]
> Any plans in future to deprecate this lib like the request?

`node-fetch` isn't going to be deprecated any time soon unless node implements fetch [natively](https://github.com/nodejs/node/issues/19393) in which case this will merly become just a temporary polyfill and a deprecation till everyone update it's node version and uses the built in version

`node-fetch` follows the spec as much as possible. So if the spec changes then we will follow.

---

I'm glad you chosen node-fetch as a http library, this will make whatever you build to also function in a browser context with little friction and no extra browserified bundles like node streams/buffers (depending on how you choose to use `node-fetch`)

**Comment by dobromyslov** [comment]
Thank you. I think AWS Lambdas and Google CloudFunctions wont update fast to the fresh Node version. Google provides stable Node 8 and Node 10 only as a beta env.

**Comment by bitinn** [comment]
I know these cloud platform love their verified production. But I didn’t realize they love it that much.

Either way, we are moving onto v3, there is a single issue on Unicode URL conversion that have prevented our release so far, once that’s solved, v3 will be out.

The deprecation for older Node LTS release’s support will still happen.

(Unless Amazon or Google are willing to donate to our project? Then perhaps we might have resources to consider their priorities.)

(From my phone)

**Comment by bitinn** [comment]
On the subject of that WHATWG URL issue, what’s everyone’s take?

Should we just include the utf8 package instead of using new URL()? I would really love to use new URL() but I don’t think nodejs core has movement to fix their issue?

(From my phone)

**Comment by gr2m** [comment]
> Is node-fetch a good substitution for the request lib

@dobromyslov creating a migration tutorial for today's `request` users looking for alternatives would be a great contribution :) If you or anyone are interested, can you create a separate issue so don't go to off-topic in this thread? A lot of people are watching it and we want to keep noise to a minimum

**Comment by Richienb** [comment]
Since #701 is now merged, we can now release `3.0.0-beta.1`.

**Comment by bitinn** [comment]
OK folks, `3.0.0-beta.1` is out, use `npm install node-fetch@next` to get it~

And `3.x` branch is merged back into `master` now, we can use it from now on.

Someone should update the README to fix the badges and point people to v3 beta releases. Thx~

**Comment by xxczaki** [comment]
Just published a short blog post, summarizing the changes made in the new release :smile:

https://medium.com/@kepinski/node-fetch-v3-beta-is-out-9fbb8091999

**Comment by bitinn** [comment]
@xxczaki cool, thx for the write-up.

on the readme update, it's probably best to distinguish `npm install node-fetch` and `npm install node-fetch@next` until v3 is stable.

also we can add a badge for the v3 branch

![next branch](https://img.shields.io/npm/v/node-fetch/next?style=flat-square)

**Comment by bitinn** [comment]
So that `3.0.0-beta.1` to `3.0.0-beta.3` definitely didn't work, I thought `pika` has been setup to work with `npm publish`, but it doesn't.

To be honest, I am surprised I cannot figure out a way to publish it properly. Someone in charge of integrating `pika` need to write me a guide or send me a PR that actually make `npm publish` work.

For now, I assume people will encounter issue with `node-fetch@next` and `node-fetch`, because it was pushed to `latest` channel without me realizing what `pika` was doing.

**Comment by Richienb** [comment]
@bitinn `npm unpublish` is able to unpublish versions published less than 24 hours ago.

**Comment by bitinn** [comment]
@Richienb `node-fetch` have too much dependency, it's not possible to `unpublish`, we can only deprecate, I have deprecated all 3.x releases.

**Comment by Richienb** [comment]
@bitinn To (hopefully) revert the `latest` tag, run `npm dist-tag add node-fetch@2.6.0 latest`

**Comment by bitinn** [comment]
@Richienb I do not believe either `npm publish` or `pika publish` works with current codebase. I tried them without modifying package.json initially, the result is invalid main as it's pointing to a non-exist index.js;

People more familiar with `pika` will need to look into it, my knowledge of this whole transpiling ecosystem stops at `babel` and `rollup`. I have always tried to avoid them when possible.

**Comment by Richienb** [comment]
@bitinn Can I try to publish it properly? Please do `npm owner add richienb node-fetch`.

**Comment by bitinn** [comment]
@Richienb if you are not running `pack publish`, please let me know the commands or steps instead. I would like to make sure I fully understand these steps, not deferring it to someone else.

when 3.x goes into stable, I will add more members to the npm ownership.

(It was on me to verify the steps with you guys, so I don't blame anyone but myself, but please give me the steps; if you need to test a solution, please fork and test it)

**Comment by Richienb** [comment]
@bitinn 

> if you need to test a solution, please fork and test it with private scope.)

Good idea!

**Comment by Richienb** [comment]
@bitinn I've just fixed the code and [verified it as working](https://www.npmjs.com/package/@richienb/node-fetch-beta-test-14-03-2020). I've set up so that you can publish `node-fetch` as if pika didn't exist.

Do a `git pull` then run `npm publish --tag next`

**Comment by jimmywarting** [comment]
just tried using `node-fetch@beta.3` in a new fresh project using node v13 with ESM module

many modules have been working fine except for `node-fetch`.

```js
import fetch from 'node-fetch'
```

```
(node:18717) ExperimentalWarning: The ESM module loader is experimental.
internal/modules/esm/resolve.js:61
  let url = moduleWrapResolve(specifier, parentURL);
            ^

Error: Cannot find main entry point for ~/project/node_modules/node-fetch/ imported from ~/project/app.js
```

---

```js
import fetch from 'node-fetch/src/index.js'
```

```
(node:18709) ExperimentalWarning: The ESM module loader is experimental.
~/project/node_modules/node-fetch/src/index.js:9
import http from 'http';
^^^^^^

SyntaxError: Cannot use import statement outside a module
```

---

I would rather have preferred to load the real deal from src and not some transpiled assets 😞

**Comment by jimmywarting** [comment]
+ require isn't avalible any longer when i have specified 


package.json
```json
{
  "type": "module",
}
```

so i have a really hard time to load node-fetch right now

**Comment by bitinn** [comment]
@Richienb I maybe wrong, but I do not believe

`const fetch = require('node-fetch');`

gives me a function.

`console.log(fetch)`

gives me:

```
{
  default: [Function: fetch] {
    isRedirect: [Function],
    Promise: [Function: Promise]
  },
  Response: [Getter],
  Headers: [Getter],
  Request: [Getter],
  FetchError: [Getter]
}
```

instead of:

```
[Function: fetch] {
  isRedirect: [Function],
  Promise: [Function: Promise],
  default: [Circular],
  Headers: [Function: Headers],
  Request: [Function: Request],
  Response: [Function: Response],
  FetchError: [Function: FetchError]
}

```

**Comment by bitinn** [comment]
Also we should fix the `npm run report` and `npm run coverage` commands, it's currently reporting 0% for me, even though all the tests passed.

**Comment by bitinn** [comment]
Some thoughts after this incidence:

- I think there are still quite a lot to do before we can consider `3.0` ready for beta. We have to dogfood our own code first, before asking others to test for us.

- But I really appreciate your efforts in this, @xxczaki @Richienb @jimmywarting 

- I dropped the ball when publishing it without dogfooding it first;

- I also shouldn't have used squash merge to pull `3.x` into `master`, losing all this history for v3 codebase.

- At this point I am happy to redo the whole merge `3.x` into `master` again by reverting `master`, fixing `3.x` to a use-able states, then `merge` or `rebase` it into `master` again.

What do you think?

**Comment by bitinn** [comment]
If we are all ok then @xxczaki @Richienb can process with this plan, I have left `3.x` codebase largely on you, so you should do the honor instead of me.

@jimmywarting and me will still do the publishing as a last line of safeguard, I hope to do it better this time around :)

**Comment by jimmywarting** [comment]
Tried beta-4 now. at least i can load fetch now but felt a bit weird that i had to do:

```js
import nodeFetch from 'node-fetch'
const fetch = nodeFetch.default
```

**Comment by NotMoni** [comment]
hey, how can i be a contributor? I would like to help.

**Comment by xxczaki** [comment]
@Richienb @bitinn @jimmywarting 

I think we can publish `3.0.0-beta.5`, as #759 got merged.

In terms of the import issues - I think the problem was occurring, because we didn't have `@pika/plugin-build-node` in our `@pika/pack` pipeline (so only the modern, ES2018 bundle was created). I added it in 5ecb48e, so everything should work fine now (although this needs to be tested).

PS: @bitinn or @Richienb, could you change my role in the node-fetch org to owner? I would like to try and fix the issues with codecov coverage reports :smile:

**Comment by xxczaki** [comment]
Also @NotMoni, sorry for a late reply. Check out #772 :smile:

**Comment by Richienb** [comment]
@xxczaki As for the bundling, we can use [conditional exports](https://nodejs.org/api/esm.html#esm_conditional_exports) to export the cjs and ESM files separately to avoid having to deal with complicated bundling.

I also changed your role to `owner`.

**Comment by NotMoni** [comment]
> Also @NotMoni, sorry for a late reply. You should check out #772 :smile:

Sure. I would love to help node-fetch

**Comment by NotMoni** [comment]
@xxczaki Hey, Count Me In. I would like to help. Awaiting the Invitaiton :D

**Comment by xxczaki** [comment]
@Richienb I tried different solutions (from a simple @babel/cli transpiling to rolling back to the old Rollup config), but nothing seems to work correctly :cry: It looks like using a default export with a couple of named exports makes things difficult for bundlers & transpilers.

**Comment by Richienb** [comment]
@xxczaki Since ESM syntax is completely incompatible with commonjs, using the previously working solution for v2 would only make the code usable if a bundler which could detect the `__esModule` property. Splitting the different types of files would make our lives much easier - see https://github.com/node-fetch/node-fetch/pull/778.

**Comment by xxczaki** [comment]
The coverage report is now working again.

**Comment by xxczaki** [comment]
@node-fetch/core

After merging #778, #777, #747 and maybe #770 and #768 we can finally release `3.0.0-beta.5`. We might also want to deprecate some of the previous beta versions, mainly due to `require('').default` issues.

**Comment by xxczaki** [comment]
https://github.com/node-fetch/node-fetch/issues/749 itself is a strong reason to deprecate the beta versions.

**Comment by JefferyHus** [comment]
Hey folks, are you still looking for some help?

**Comment by xxczaki** [comment]
Ok, so the 3 PRs are now merged. In terms of the other 2 (#770 and #768) - I'm not sure whether we want to include them in the `beta.5` release or not. It looks like at least one of them includes a breaking change.

cc @bitinn @Richienb @jimmywarting

**Comment by xxczaki** [comment]
@JefferyHus Please check #772 and if you are interested - reply to that issue. This one is watched by people, who just want to stay updated about the upcoming `v3` release :smile:

**Comment by bitinn** [comment]
> we can finally release 3.0.0-beta.5. We might also want to deprecate some of the previous beta versions, mainly due to require('').default issues.

@xxczaki yeah, when you publish the next beta version you should deprecate the old ones.

**Comment by xxczaki** [comment]
Ok, I just published `3.0.0-beta.5` :smile:

- [Changelog](https://github.com/node-fetch/node-fetch/blob/master/docs/CHANGELOG.md#v300-beta5)

All previous beta versions are deprecated.

**Comment by NotMoni** [comment]
Can we convert res to JSON by default if possible?

**Comment by NotMoni** [comment]
We also should move to GitHub action in favor of TravisCI. this means we can manage everything in one place.

**Comment by tinovyatkin** [comment]
Did you consider dropping feature that allows overwriting Promise implementation? Without that change, you can't use native async function in code, or event return a Promise from an external dependency (for example, the Blob implementation) without wrapping it into fetch.Promise. 
It looks like a well-outdated feature that produces lot of troubles without any real benefit.

**Comment by bitinn** [comment]
@tinovyatkin I am open to this idea, AFAIK nodejs 10's native promise is fast enough.

I guess my only concern is how many people out there are using this to hook more debugging or analytics into node-fetch's response object; it's likely an edge case usage though.

(But I shouldn't be the sole decider, I will let org members decide.)

**Comment by xxczaki** [comment]
I'm :+1: on dropping this. We will obviously need to document this.

**Comment by tinovyatkin** [comment]
@bitinn Do you want me to submit a PR showcasing the benefits of dropping custom Promises (async refactoring, etc) or wait for the team consensus first?

**Comment by bitinn** [comment]
@tinovyatkin if you have got the time, we definitely welcome a PR with benefit showcase.

**Comment by xxczaki** [comment]
I think we can release `3.0.0-beta.6` after doing the following tasks:

- [ ] Merge #810
- [ ] Merge #785
- [ ] Drop custom Promises (AFAIK PR in progress)

Maybe also #799, but I feel like this needs to be further discussed.

**Comment by JefferyHus** [comment]
I will be able to help on this `Drop custom Promises`

**Comment by NotMoni** [comment]
Issue and PR templates should be made before v3.

**Comment by gr2m** [comment]
I think Issue/PR templates can be introduced at any time, they are not a feature relevant to v3.

That being said, would you like the new Discussions tab? It helps to keep issues to be true bug reports, and move everything else to discussions. E.g. check out https://github.com/zeit/next.js/issues/new/choose. Let me know, I can try to sign us up. I can setup the issue templates, too

**Comment by NotMoni** [comment]
@gr2m I already made a pr with the templates.

But, I do think the discussions tab will be very helpful.

**Comment by xxczaki** [comment]
@tinovyatkin Shall I release `3.0.0-beta.6` now, of we should wait and drop custom promise?

**Comment by tinovyatkin** [comment]
@xxczaki I would love to see `beta.6` released now as it closes so many issues already, most importantly with TypeScript declarations and Istanbul instrumentation in code.
I actually think that beta 5 must be immediately deprecated due to these problems.
Due to #805 `node-fetch` will be true Node dual-system module and it's good to receive as much feedback from userland as possible.
I was assuming that @JefferyHus is working on `async` refactoring, but not seeing any PR in progress, should I take it?

**Comment by xxczaki** [comment]
`node-fetch@3.0.0-beta.6` published! All previous beta versions are now deprecated.

**Comment by tinovyatkin** [comment]
I believe that problems fixed by #841 and #845 are enough to release the next beta when these PRs will be merged (as both contains/mitigating breaking changes)

**Comment by xxczaki** [comment]
@tinovyatkin Shall I release `3.0.0-beta.7` now or we want to get some other PRs merged (like #842, #843 or #846)?

**Comment by tinovyatkin** [comment]
#846 fixes one old small bug with partial responses on Node < 12.10, that worth waiting to be merged. The rest of PRs are internal changes and can be included later....

**Comment by tinovyatkin** [comment]
#853 that fixed bug in data URI handling would be nice to include too...

**Comment by SomaticIT** [comment]
Any news about the `3.0.0-beta.7`?

**Comment by xxczaki** [comment]
@SomaticIT We are waiting for https://github.com/node-fetch/node-fetch/pull/846 to be merged :smile:

**Comment by tinovyatkin** [comment]
@xxczaki I believe we are ready to go with beta 7:
- The bug from #866 is more about spec compliance and seems to have limited real-life use, so, can be merged later
- #842 is just an internal refactoring that's nice to have but can wait
- #603 still have Windows test failing and looks like work in progress

**Comment by Richienb** [comment]
Since https://github.com/node-fetch/node-fetch/pull/866 is targeting a very specific part of the fetch API and is trying to make it more compliant, we likely want to include it in beta 7.

**Comment by tinovyatkin** [comment]
@Richienb probably it's worth releasing a new version of `fetch-blob` (with changes by @jimmywarting and include new typing from https://github.com/node-fetch/fetch-blob/pull/45 to decouple `dom.lib`) and to include the upgraded dependency to `beta.7`?

**Comment by octet-stream** [comment]
> #603 still have Windows test failing and looks like a work in progress

I think it's done, unless you will request some changes. I've just added more tests for form-data encoding helpers.

**Comment by xxczaki** [comment]
Update: `3.0.0-beta.7` released.

**Comment by gr2m** [comment]
I've found that `timeout` was still documented in the README, I started a PR at #877. 

While investigating, I found other options that are defined in the [`RequestInit`](https://github.com/node-fetch/node-fetch/blob/1fdc218a64f8bc2c85bec0ca59a85973ef5e1f0a/%40types/index.d.ts#L56-L88) but are not documented as node extensions:

- counter
- hostname
- port
- protocol

can we add them to the node-only extensions?

**Comment by ryanflorence** [comment]
Any plans for `await request.formData()` ? Been cruising through issues and can't find anything.

**Comment by tinovyatkin** [comment]
Now, when the current LTS is 14.x, I believe it's reasonable to drop `< 12.18` support in the `v3` release? We can drop quite a few legacy logic then...

**Comment by tekwiz** [comment]
@tinovyatkin 12.x is still under maintenance support for another 16 months (and 10.x is also for another 4 months). I'm afraid that if we drop support for <12.18, we will just end up with a bunch of irritated users with lots of "wontfix" issues. If the intent is just to remove the logic branches, I think the public relations benefit of leaving them is greater than the cost.

**Comment by gr2m** [comment]
I agree with @mariusa. v2 has been around for a long time and is stable. We can continue to accept bug fixes for v2 for a while.

With v3, we should drop support for Node < 12.8 and build node-fetch as native ES Module, without any transpilation to CJS, just make a clear break, similar to what Sindre laid out here: https://blog.sindresorhus.com/get-ready-for-esm-aa53530b3f77

**Comment by martinheidegger** [comment]
I know this has been closed for a while, but would you be open to add HTTP_PROXY/HTTPS_PROXY support to node-fetch? #195

**Comment by xxczaki** [comment]
I'm back and happy to make a PR to use pure ESM :smile:

**Comment by gr2m** [comment]
Looks like @Richienb and @jimmywarting are in favor, too

Any thoughts @bitinn?

**Comment by bitinn** [comment]
I am onboard, @xxczaki you can do it.

My hope is v3 stable gets out this year :)

Also note there are some outstanding PRs, fixing known issues is probably more useful than a rewrite.

(but I understand the desire to have a clean break.)

**Comment by xxczaki** [comment]
@node-fetch/core @node-fetch/reviewers and others,

I opened a pull request which includes the ESM transition and more. Feedback is appreciated, as merging this will allow for the release of `3.0.0-beta.10` which will include some important bug fixes.

**Comment by jimmywarting** [comment]
the v3 beta have been around for quite a while now, can't make a release happen soon?

**Comment by gr2m** [comment]
I'm sorry I didn't have the time to review @xxczaki pull request for the ESM transition yet. I do have it on top of my work todo list though, I'll very likely get to it this coming week.

**Comment by jimmywarting** [comment]
Good news! 

WHATWG streams have landed in NodeJS (as experimental)
https://github.com/nodejs/node/pull/39062

Think we should start using it whenever possible, might be worth looking into adding whatwg stream polyfill now!
It would be a major breaking change. Maybe start planing for v4 straight away

**Comment by jimmywarting** [comment]
ESM has landed! 🚀 

This issue have become quite long to scan through.
Is it worth shipping `3.0.0-beta.10` now, or should we fix any other issues?

Going from `beta.9` to current version is a major breaking change from going from cjs to esm that i feel that we can safely release v3 without having to make any additional breaking changes later (except when we start using real whatwg streams - but will probably hold of from this a tiny bit longer). 
Feature PR will mostly only need minor/patch fixes.

One thing that worries me now is that ppl that are already using 3.0.0-beta.9 with `require()` would start breaking if we released `3.0.0-beta.10`? maybe should just go for `3.0.0`?

**Comment by gr2m** [comment]
> One thing that worries me now is that ppl that are already using 3.0.0-beta.9 with `require()` would start breaking if we released `3.0.0-beta.10`?

breaking changes are ok for pre-releases such as `-beta.X`, I wouldn't worry about it. People who have `3.0.0-beta.9` installed will not automatically get `3.0.0-beta.10` when they run `npm install`, it needs an explicit change in the `package.json`.

**Comment by jimmywarting** [comment]
> People who have `3.0.0-beta.9` installed will not automatically get `3.0.0-beta.10` when they run `npm install`

Good to know. Should we publish the next beta version? 

After that I say we hold off for a while with any new features/PR and only fix bugs that might arise.
If there is no major issue I think that it has gotten time that we start releasing v3 - it have been put off for such a long time now and there are many new feature/bug fixes since v2.

(unless someone else _really_ wants something else to be fixed before we ship v3)

**Comment by jimmywarting** [comment]
beta.10 release!

Entering maintenance mode for 2-4 week 🙂 
let's see how it plays out. and maybe we can release it as stable afterwards

**Comment by styfle** [comment]
It looks like the size increased quite a bit

- node-fetch@3.0.0-beta.9 [![node-fetch@3.0.0-beta.9](https://packagephobia.com/badge?p=node-fetch@3.0.0-beta.9)](https://packagephobia.com/result?p=node-fetch@3.0.0-beta.9)
- node-fetch@3.0.0-beta.10 [![node-fetch@3.0.0-beta.10](https://packagephobia.com/badge?p=node-fetch@3.0.0-beta.10)](https://packagephobia.com/result?p=node-fetch@3.0.0-beta.10)

Seems like most of that is from [web-streams-polyfill](https://packagephobia.com/result?p=web-streams-polyfill).
Is that needed for node-fetch?

**Comment by jimmywarting** [comment]
@styfle 

b/c NodeJS added native support for `whatwg:streams` we thought: Ok, so whatwg streams are actually going to become useful in NodeJS landscape.
So i was thinking: Ok so if we are going to be a references a nodejs implementation then we **have** to do everything nodejs is doing so there isn't any differences between implementations

so: fetch-blob was the first to adapt web-streams-polyfill
now i'm also thinking that `Body.body` should also be a whatwg:streams in some feature release, so yea, we are going to depend on it from now on. 

NodeJS is also going to ship fetch in core one day. and we wish to be the polyfill that references nodejs implementation so it can be backward compatible until everyone starts using nodes (upcoming) built in fetch impl that means we should also be using whatwg stream for compatibility reasons.

#1217 tracks the size issue, and @MattiasBuelens is working on reducing the size here: https://github.com/MattiasBuelens/web-streams-polyfill/pull/83
